### PR TITLE
Migrate to optimistic transaction db

### LINKED ISSUE #2060: Use WriteBatch for state update atomicity on processing records
For example in the TypedStreamProcessor create an batch and process is executed inside this batch.

We need to adjust the Zb-db implementation such that nested batches are recognized and reuses the already open batch object.

### LINKED ISSUE #2061: Introduce the OptimisticTransactionDb to achieve atomicity in stream processing
Use the OptimisticTransactionDb (RocksDB) to achieve atomicity in stream processing, this should replace the usage of the WriteBatches, which was introduces with #2060

### PR DESCRIPTION
**Pull request by ChrisKujawa**
* migrate `ZeebeDb` impl from `RocksDB` to `OptimisticTransactionDb` (extended `RocksDB`)
 * fix failing tests and race conditions
 * use transaction aware iterators
 * rename batch to transaction, since also API should give transaction
 like guarantees
 * process event is now executed inside an transaction

Currently all calls at the `ZeebeDB` will end in a transaction, which means if there is no open transaction a new will be created. This was as a first step easier to implement and should not be a problem, since processing is now executed in a transaction anyway, so there is no difference (no new transaction will be created for nested transactions).
If we want to change this, we need to do some more reflection to expose the `RocksDB` methods, since we can't extend the `OptimisticTransactionDB`.

closes #2061 
closes #2060

### CODE REVIEW

**Review by npepinpe — COMMENTED** [review]
Just wondering, why change the wording of RocksDB everywhere for `TransactionDb`, e.g. `ZeebeRocksDb` -> `ZeebeTransactionDb`, if anyway we don't keep the old implementation? Why the distinction?

**Comment by Zelldon** [comment]
I started as first with transaction impl on side with the old implementation. After finishing this I decided to remove the old one, but was not sure if we need it or not. The naming of the impl can't be seen from outside anyway so I think this is not a real issue.

File: zb-db/src/main/java/io/zeebe/db/impl/rocksdb/transaction/ZeebeTransactionDb.java [hunk 1]
```diff
@@ -143,49 +128,56 @@ protected long getColumnFamilyHandle(ColumnFamilyNames columnFamily) {
...
-      throw new RuntimeException("Unexpected error occurred trying to write RocksDB entry", e);
-    }
+          try {
+            currentTransaction.put(
+                columnFamilyHandle,
+                keyBuffer.byteArray(),
+                key.getLength(),
+                valueBuffer.byteArray(),
+                value.getLength());
+          } catch (Exception e) {
```
Comment (npepinpe) [inline zb-db/src/main/java/io/zeebe/db/impl/rocksdb/transaction/ZeebeTransactionDb.java]: Can we remove all those catch statements and have a single one in `transaction`? We're anyway catching everything there.

**Review by npepinpe — APPROVED** [review]
Looks good, disappointing that we can't extend `OptimisticTransactionDB` though. It's a weird choice they made considering the C++ API is extensible...

Only comment is more of a cleanup thing (since we wrap exceptions twice I guess?), I wouldn't consider a blocker.

**Comment by Zelldon** [comment]
bors retry

**Pull request merged**